utility/FullIo.cpp

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2011-2014, Intel Corporation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation and/or
+ * other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "FullIo.hpp"
+
+#include <cerrno>
+#include <unistd.h>
+
+namespace utility
+{
+
+/** Add a pointer and */
+template <class Ptr>
+Ptr add(Ptr ptr, size_t count) {
+    return (char *)ptr + count;
+}
+
+template <bool allowEmpty, class Buff>
+static bool fullAccess(ssize_t (&accessor)(int, Buff, size_t),
+                       int fd, Buff buf, size_t count) {
+    size_t toAccess = count; // Bytes left to access
+    Buff endBuf = add(buf, count); // Pointer to the end of the buffer
+    while (toAccess > 0) {
+        ssize_t acceded = accessor(fd, add(endBuf, -toAccess), toAccess);
```

Comment:
Are you sure the code is working here ? for me -toAccess is giving you a huge value as you put a signed negative value in a size_t (but maybe I am wrong)
Maybe you can accumulate valid acceded value and use it as index (and verifying there is no overflow) It will avoid the soustraction